pkg/api/subscribeWorker.go

```diff
@@ -0,0 +1,196 @@
+package api
+
+import (
+	"context"
+	"database/sql"
+	"encoding/hex"
+	"fmt"
+	"time"
+
+	"github.com/xmtp/xmtpd/pkg/db"
+	"github.com/xmtp/xmtpd/pkg/db/queries"
+	"github.com/xmtp/xmtpd/pkg/proto/xmtpv4/message_api"
+	"go.uber.org/zap"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	subscriptionBufferSize    = 1024
+	maxSubscriptionsPerClient = 10000
+	SubscribeWorkerPollTime   = 100 * time.Millisecond
+	subscribeWorkerPollRows   = 10000
+	maxTopicLength            = 128
+)
+
+type listener = chan<- []*message_api.OriginatorEnvelope
+
+// A worker that listens for new envelopes in the DB and sends them to subscribers
+// Assumes that there are many listeners - non-blocking updates are sent on buffered channels
+// and may be dropped if full
+type subscribeWorker struct {
+	ctx context.Context
+	log *zap.Logger
+
+	dbSubscription <-chan []queries.GatewayEnvelope
+	// Assumption: listeners cannot be in multiple slices
+	globalListeners     []listener
+	originatorListeners map[uint32][]listener
+	topicListeners      map[string][]listener
+}
+
+func startSubscribeWorker(
+	ctx context.Context,
+	log *zap.Logger,
+	store *sql.DB,
+) (*subscribeWorker, error) {
+	log = log.Named("subscribeWorker")
+	q := queries.New(store)
+	pollableQuery := func(ctx context.Context, lastSeen db.VectorClock, numRows int32) ([]queries.GatewayEnvelope, db.VectorClock, error) {
+		envs, err := q.
+			SelectGatewayEnvelopes(
+				ctx,
+				*db.SetVectorClock(&queries.SelectGatewayEnvelopesParams{}, lastSeen),
+			)
+		if err != nil {
+			return nil, lastSeen, err
+		}
+		for _, env := range envs {
+			// TODO(rich) Handle out-of-order envelopes
+			lastSeen[uint32(env.OriginatorNodeID)] = uint64(env.OriginatorSequenceID)
+		}
+		return envs, lastSeen, nil
+	}
+
+	vc, err := q.SelectVectorClock(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	subscription := db.NewDBSubscription(
+		ctx,
+		log,
+		pollableQuery,
+		db.ToVectorClock(vc),
+		db.PollingOptions{
+			Interval: SubscribeWorkerPollTime,
+			NumRows:  subscribeWorkerPollRows,
+		},
+	)
+	dbChan, err := subscription.Start()
+	if err != nil {
+		return nil, err
+	}
+	worker := &subscribeWorker{
+		ctx:                 ctx,
+		log:                 log,
+		dbSubscription:      dbChan,
+		globalListeners:     make([]listener, 0),
+		originatorListeners: make(map[uint32][]listener),
+		topicListeners:      make(map[string][]listener),
+	}
+
+	go worker.start()
+
+	return worker, nil
+}
+
+func (s *subscribeWorker) start() {
+	for {
+		select {
+		case <-s.ctx.Done():
+			return
+		case new_batch := <-s.dbSubscription:
```

Comment:
I am a bit uncomfortable that we are passing a slice to a channel. In this simple use case its correct, but slices are mutable and not thread safe. So it makes me not trust this code.

Reply:
Can you be more specific about the case you are worried about? In general I'm not sure if there's a hard-and-fast rule that all objects passed into a channel must be immutable (although it would be suspect if they were mutated while they were in the channel, which we don't do)